### PR TITLE
Docs: add a living faq document for sharing dev solutions

### PR DESCRIPTION
**Pull request by qwint**
## What is this fixing or adding?
https://discord.com/channels/731205301247803413/1214608557077700720/1229567644303101973

### CODE REVIEW

**Comment by NewSoupVi** [comment]
I think the existence of this doc is good, although I think if it gets merged I will immediately PR some additions to the current listed cases :D

I would consider using "world" in the file name instead of "apworld" for now, but I don't feel too strongly about it. 
Reasoning: Right now apworld still has an implication of `.apworld` specifically where world just means "anything that can be run somewhere on Archipelago", Although the meaning of "apworld" is shifting a bit to just mean what "world" used to mean, which is fine, as that is the long term goal

**Comment by Silvris** [comment]
So should feedback on what should be included take place here, or should there be a more dedicated discussion in the Discord (for this first iteration at least)?

**Comment by qwint** [comment]
> So should feedback on what should be included take place here, or should there be a more dedicated discussion in the Discord (for this first iteration at least)?

personally I'd like this to be a living document so I'd rather it be merged and prs opened to add things, but I'm willing to take suggestions of more things to add to show _how_ the document should be filled (i.e. taking Berserker's suggestion and showing an answer that links back to the relevant core file)

File: docs/apworld dev faq.md
Comment (beauxq) [inline docs/apworld dev faq.md:8]: I don't know what it looks like in a markdown renderer, but this looks a little strange in this plain text form.

The `##` header is closer to the unrelated text above it, than it is to the related text below it.

I think it would make more sense to have 1 empty line between the `##` header and its corresponding text below it, and 2 empty lines above the header.

You could consider adding `---` line breaks between items too.

markdown is designed to look good in plain text as well as having a good rendered form, so I'd like to make sure it also looks good in plain text form.

Comment (alwaysintreble) [reply docs/apworld dev faq.md]: should be one line between on both sides https://google.github.io/styleguide/docguide/style.html#headings there should be breaks between code too.

File: docs/apworld dev faq.md
Comment (alwaysintreble) [inline docs/apworld dev faq.md:19]: this link should link to the actual relevant part not the entire file. also due to the [sphinx docs migration](https://github.com/ArchipelagoMW/Archipelago/pull/2909) it should be a direct github link

File: docs/apworld dev faq.md
Comment (qwint) [reply docs/apworld dev faq.md]: liked the `---` idea but i'm not sure if that supplanted the double newline idea or not, so i moved forward with just one empty newline between every part but adding the `---` between sections (wrapped with their own newlines)
also if you want to see the markdown render in github from the diff, you can "view file" from the overflow menu and it renders .md files

File: docs/apworld dev faq.md
Comment (qwint) [reply docs/apworld dev faq.md:19]: I didn't want to link to a line number (because they could change and it would be a horrid mess to maintain that) and couldn't figure out to link otherwise, do you have an example?

File: docs/apworld dev faq.md
Comment (qwint) [reply docs/apworld dev faq.md]: > should be one line between on both sides https://google.github.io/styleguide/docguide/style.html#headings there should be breaks between code too.

did not see this before updating/replying, this is agreeing with my current spacings right? 
i'm not certain what you mean about breaks between code though, like change the first code block to be this?:
```
early_item_name = "Sword"

self.multiworld.local_early_items[self.player][early_item_name] = 1
```

File: docs/apworld dev faq.md
Comment (qwint) [reply docs/apworld dev faq.md:27]: that note should probably be there regardless; added

File: docs/apworld dev faq.md
Comment (ScipioWright) [inline docs/apworld dev faq.md:3]:
```suggestion
This document is meant as a reference tool to show the best known solutions to common problems when developing an apworld.
```

**Review by ScipioWright — APPROVED** [review]
I think this is overall a great thing to add in, and will try to note down some items to add in once this gets merged.

**Comment by NewSoupVi** [comment]
Relevant discussion about what my "approval criteria" is for this PR:
https://discord.com/channels/731205301247803413/731214280439103580/1248454687103909898

**Review by NewSoupVi — CHANGES_REQUESTED** [review]
Willing to merge (after reading it over) if a line is added to `CODEOWNERS` to signal who is going to be responsible:tm: for this document. Qwint and Scipio have volunteered, they may try to recruit other people as well if they wish of course.

**Review by agilbert1412 — APPROVED** [review]

File: docs/apworld dev faq.md
Comment (beauxq) [inline docs/apworld dev faq.md:23]: I don't like advertising this solution in a document like this.

I'm not saying that this code is bad or that worlds shouldn't be doing this - just that it's not ideal enough to be advertising it in this document.

This looks like a workaround for having implemented something else in a less-than-ideal way.

In general I don't like the pattern of world X asking the multiworld what world X did (asking "How many locations did I create?")
Ideally, world X should already know what world X did because world X is world X.

This might be the best workaround after having made lots of other choices that led to this situation.
But the problem is that people will read this, seeing it as an ideal solution, before they have made those other choices that led to this situation.

Comment (ScipioWright) [reply docs/apworld dev faq.md:23]: Personally, I don't see just asking the multiworld how many locations are unfilled as being an issue.
But, if you do see an issue, please suggest a specific change here.

If the standard for the document is "only the most ideal solution", then we might as well just close this PR now, imo. No one will ever agree on the most ideal solution for most things that would be useful to have in a doc like this.

Comment (beauxq) [reply docs/apworld dev faq.md:23]: > If the standard for the document is "only the most ideal solution"

This is a misrepresentation of what I'm saying and a strawman.
"it's not ideal enough" does not mean, imply, or suggest "only the most ideal"

Comment (qwint) [reply docs/apworld dev faq.md:23]: the issue states "I have multiple settings that change the item/location pool counts and need to balance them out"
so it's already under the assumption that they have made those design decisions.
But we could always add some extra note/nuance to the header along the lines of "while it is recommended that you build out your locations/items so that the world knows how many locations it created, You can create extra filler based on . . ." 

that aside, some rambles / nitpicks with your argument that essentially amount to: we should probably have a better statement of goal in the document,

saying it's not good enough to be in this document is weird when the point of the document is to have referenceable answers to questions that come up (and this question does come up),
I don't see this snippit as prescriptive - i don't expect everyone to make a new create_non_filler_items() method after reading this. And, albiet its a lot more obvious of an example but i also don't expect everyone to submit "Sword" as an early item.
so showing _a_ way to get location count and more importantly showing them how to use it efficiently makes sense to me

also the idea that a world should know how many locations it submitted and not need to ask the multiworld is strange to me because multiworld.get_locations() (and thus get_unfilled_locations()) already use a cache, so while yes, conceptually its weird to forget what you've already done, it also is weird to expect each individual World to make a better caching system (or alternatively, expect this one FAQ answer to define 1/3 of everyone's World code,,)

Comment (beauxq) [reply docs/apworld dev faq.md:23]: I wonder how many people are going to find this document when they're starting their world development, and how many people will point to this document for someone just starting their world development,
compared to pointing to / finding this document after a specific question comes up.

If it's more of the latter, I think that makes this example not as much of a problem.
But I don't know which it will be.

My concern is that showing this example to someone who hasn't started making decisions yet about how to create their locations and items, could lead them to making worse decisions about how to create their locations and items.

The specific information used in this example (the number of unfilled locations) is not cached by the multiworld, so this has to take extra time checking to see whether the locations are filled.
(I would think that this was a very good example to show new world developers if it were common for other worlds to put items in my locations before I create my items for the item pool. But that isn't a thing.)

Comment (ScipioWright) [reply docs/apworld dev faq.md:23]: > > If the standard for the document is "only the most ideal solution"
> 
> This is a misrepresentation of what I'm saying and a strawman. "it's not ideal enough" does not mean, imply, or suggest "only the most ideal solution"

Please don't call my argument a "strawman". That was not the intent. We should assume good faith in each other, or we should just not interact.

But yes, the wording was a little exaggerated, but my point falls in line with qwint's: This is a fine way to get your unfilled locations, and you even stated as such. The amount of time taken to call `get_unfilled_locations` is probably negligible in a multiworld gen (please correct me if I'm wrong here). Adding some nuance to it, like qwint stated, would be a good idea here. If you have a suggestion for that wording, please share that.

**Comment by qwint** [comment]
> Willing to merge (after reading it over) if a line is added to `CODEOWNERS` to signal who is going to be responsible™ for this document. Qwint [sic] and Scipio have volunteered, they may try to recruit other people as well if they wish of course.

I asked in the Discord, but I'll re-ask here for completeness:
What wording would have to be added to `CODEOWNERS` to represent the header for owners/maintainers for particular Docs and what those responsibilities entail

File: docs/apworld dev faq.md
Comment (beauxq) [reply docs/apworld dev faq.md:23]: > Adding some nuance to it, like qwint stated, would be a good idea here. If you have a suggestion for that wording, please share that.

Maybe something like this:
In an ideal situation your system for producing locations and items wouldn't leave any opportunity for them to be unbalanced. But in real, complex situations, that might be unfeasible.

File: docs/CODEOWNERS
Comment (beauxq) [inline docs/CODEOWNERS:203]:
```suggestion
/docs/apworld\ dev\ faq.md @qwint @ScipioWright
```
I think white space needs escape.

Comment (qwint) [reply docs/CODEOWNERS:203]: my alternative solution i didn't get to yet was just to change them to underscores in the doc and the filename, i think that'd be better especially for readability

**Review by ScipioWright — APPROVED** [review]

File: docs/apworld dev faq.md
Comment (qwint) [reply docs/apworld dev faq.md]: found a filter that seems to work, but not always, and was noted as not supported by all browsers
I think I'll leave it with the filter because worst case it functions the same way as it did without the filter

Also made it a direct link, though once we get the sphinx docs I imagine we may want to link to those docs instead of the source code,

**Review by Exempt-Medic — APPROVED** [review]
Ideally I'd like some fuller answers to these questions (like mentioning placing items yourself and adding more locations, or showing a `for` loop instead of a `while`) but these are fine as-is and I could always PR that afterwards

**Review by NewSoupVi — APPROVED** [review]

**Pull request merged**
